common/src/dataloader/cache.rs

```diff
@@ -16,62 +19,70 @@ pub trait Cache<L: Loader<S>, S = RandomState> {
     }
 }
 
-/// # Safety
-///
-/// This trait is marked as unsafe because the implementor must ensure that the Cache is safe for concurrent access.
-/// This will almost always be with some kind of interior mutability. Such as a `RwLock` or `Mutex`. Or if the cache performs no-ops on mutation.
-/// Look at `SharedCache` for an example.
-pub unsafe trait AutoImplCacheRef<L: Loader<S>, S = RandomState>:
-    AutoImplCacheMutRef<L, S>
-{
+#[repr(transparent)]
+pub struct EmptyDerefMut<T>(T);
+
+impl<T> Deref for EmptyDerefMut<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T> DerefMut for EmptyDerefMut<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
 }
 
-#[inline(always)]
-#[allow(clippy::mut_from_ref)]
-#[allow(invalid_reference_casting)]
-fn upcast<T: AutoImplCacheRef<L, S>, L: Loader<S>, S>(cache: &T) -> impl Cache<L, S> {
-    // Safety:
-    // This is safe because the trait `AutoImplCacheRef` is marked as unsafe and therefore the implementor must ensure that the
-    // Cache is safe for concurrent access. This is used to implement cache for &T where T is a Cache.
-    // The issue is we need to upcast the reference to a mutable reference, even though the implementor only ever requires a reference.
-    // This is not safe unless T has some kind of interior mutability.
-    unsafe { &mut *(cache as *const T as *mut T) }
```

Comment:
There was actually nothing wrong with the way we used this. The issue would have been
```rs
struct CImpl {
   cache: Mutex<i32>
}

impl Cache {
  fn clear(&mut self) {
      *self.cache.get_mut() = 0;
  }
   ...
}
```

If the caller implemented the unsafe trait incorrectly it would be UB; now you might say well that is the point of the unsafe part. and you would be right, but i might as well catch it before they fuck up.